### ISSUE TITLE
feature: expose "orphan" propagation policy in ks binding object

### ISSUE THREAD

**Issue by clubanderson** [kubestellar/kubestellar#1839]
### Feature Description

KS syncer 0.14 and older created and overwrote objects on remote clusters, but did not delete them. 0.20 creates, updates, and deletes objects when manifestwork is created, updated, and deleted. This is nice, but is dangerous in many cases where you accidentally take over clusterscoped objects using too widely a specified predicate. 

### Proposed Solution

expose the 

deleteOption:
    propagationPolicy: 

somehow in our binding policy

### Want to contribute?

- [ ] I would like to work on this issue.

**Comment by pdettori** [comment]
It appears that without this control it is possible to overtake clusterroles such as cluster-admin and namespaces such as cube-system and accidentaly deleting them, making effectively the WEC inoperable. ManifestWork does have these controls but we do not expose them and we use the default foreground delete behavior.

See https://open-cluster-management.io/concepts/manifestwork/

```
Delete options
User can explicitly choose not to garbage collect the applied resources when a ManifestWork is deleted. The user should specify the deleteOption in the ManifestWork. By default, deleteOption is set as Foreground which means the applied resources on the spoke will be deleted with the removal of ManifestWork. User can set it to Orphan so the applied resources will not be deleted. 
```

**Comment by pdettori** [comment]
There is also a `SelectivelyOrphan` option - perhaps a set of default resources that should never be deleted should be also something we could always inject.

**Comment by nirrozenbaum** [comment]
making sure hub doesn’t overtake cluster admin or kube system resources should not be solved with these flags, it should be solved with RBAC.
today there is nothing in ks that handles permissions of users and that’s the main issue here (user is allowed to override kube system resources for example).
somewhere in March I plan on putting on the table an initial design for inventory which includes aspects of rbac like impersonation. 
having said that, rbac is a completely separate topic and should be handled separately, so my design will work correctly with rbac iff the roles are somehow synced between hub and wecs.

@clubanderson getting back to the described issue, can you describe the requirements without referencing 0.14? we were not part of that work and we don’t know all the details about how things worked. something like “as a user, I would like to..” with explanation on the what and why.

**Comment by ezrasilvera** [comment]
@nirrozenbaum RBAC will not help here as we want to control access (deletion/update) for specific object (not a resource type)

**Comment by nirrozenbaum** [comment]
I think a better definition here is needed.
let's put aside KubeStellar and think about a single cluster.
how do you decide which user can update/delete `cluster-admin` clusterrole, objects in `kube-system` namespace, etc (all the sensitive objects)?
the answer is - whoever has access to do it (RBAC). 

in KubeStellar, there is additional handling needed due to multi cluster hierarchy (hub layer + wecs layer), but I have no doubt RBAC is one aspect of the solution to this problem.

**Comment by nirrozenbaum** [comment]
for example - let’s assume hub has a special clusterrole with admin permission that doesn’t exist in the wec. can a user distribute this clusterrole to wecs and by this become admin on the wec? (not related to overtaking existing objects in the wec).
RBAC should be used to prevent such things

**Comment by ezrasilvera** [comment]
@pdettori @clubanderson I don't think the issue here is "exposing orphan policy"  this is a solution proposal and we first need to clearly define the requirement and use case we want to support. 
One of the initial questions we need to ask is: Should KS be able to manage (update/delete) objects that were not deployed through KS. 
Once we agree we want to disallow the option to update "foreign" objects there are several ways to enforce that both on the HUB and/or on the WEC.
This can also include steps to limit the selection of objects on the hub (@nirrozenbaum suggestions is a first step in that direction) .

**Comment by MikeSpreitzer** [comment]
> KS syncer 0.14 and older created and overwrote objects on remote clusters, but did not delete them.

I think this is not accurate. For _Namespace_ objects that was true; for other kinds of objects, the syncer would delete them when the KubeStellar user stopped asserting their presence in the WEC.

**Comment by MikeSpreitzer** [comment]
I agree that the concern here overlaps with the need for better authorization control (RBAC is just one of the forms of authorization in Kubernetes, which itself is just part of the access control story). But I do not think we can solve the whole problem with RBAC. I think we have the following requirement:

As a user of KubeStellar, I want to be able to use KubeStellar to add content to objects that other parties want to exist on my WECs. Examples include labels on namespaces, and Secrets connected to ServiceAccounts (NOT the automatically managed Secrets; remember that ServiceAccount ALSO allows user-managed Secret associations).

**Comment by nirrozenbaum** [comment]
IMO this requirement is not well defined.
who is the source of truth? hub or wec?
for example, wec uses label 'test: xxx` but hub wants `test:yyy`.
